More than polarization raster in a single HDF5

LiangJYu opened the pull request:
This PR modifies s1_geocode_slc so that bursts with the same burst ID but different polarizations are stored in the same HDF5.

LiangJYu commented on src/compass/s1_geocode_slc.py:
```diff
@@ -24,6 +24,22 @@
+    # Dict to group bursts with the same burst ID but different polarizations
+    # key: burst ID, value: list[S1BurstSlc]
+    x = {}
+    for b in bursts:
+        burst_id_str = str(b.burst_id)
+        if burst_id_str not in x:
+            x[burst_id_str] = burst
+        else:
+            x[burst_id_str].append(burst)
+
+    for k, v in x.items():
+        yield k, v
```
Cleaner with following pattern:
```
In [29]: def _bursts_grouping_generator(bursts):
    ...:     g =  itertools.groupby(bursts, key=lambda b: str(b.burst_id))
    ...:     for k, v in g:
    ...:         yield k, list(v)
    ...:

In [30]: list(_bursts_grouping_generator(bursts))
Out[30]:
[('t087_185683_iw1',
  [Sentinel1BurstSlc(burst_id=t087_185683_iw1),
   Sentinel1BurstSlc(burst_id=t087_185683_iw1)])]
```

vbrancat commented on src/compass/s1_geocode_slc.py:
```diff
@@ -139,28 +124,56 @@ def run(cfg: GeoRunConfig):
+                pol = b.polarization
+
+                # Load the input burst SLC
+                temp_slc_path = f'{scratch_path}/{out_paths.file_name_stem}_{pol}_temp.vrt'
```
Can we have a way to automatically specify the output path ?

LiangJYu replied:
To an extent, [runconfig.create_output_paths](https://github.com/opera-adt/COMPASS/blob/760b1b2cab082053febed69467a51ddd35d60e26/src/compass/utils/runconfig.py#L276) does this. Were you thinking about a tempfile? Or would you prefer `create_output_paths` do more?

vbrancat commented on src/compass/s1_geocode_slc.py:
```diff
@@ -139,28 +124,56 @@ def run(cfg: GeoRunConfig):
+                                          native_doppler,
+                                          image_grid_doppler, ellipsoid, threshold,
+                                          iters, blocksize, flatten,
+                                          azimuth_carrier=az_carrier_poly2d)
```
Can you please update this branch and have it in sync with the main repository? I do not see the LUT allocated here

vbrancat reviewed (commented):
Thank you @LiangJYu. I tested the PR and there are two issues:

1. I fixed the polarization in the runconfiguration file to `dual-pol`. However, the generated products still contain only the VV channel
2. The filename of the generated product contains the polarization channel (e.g., VV). I think we should remove it from the output file name for clarity. The metadata still will have a flag to indicate the number of polarizations.
3. The branch seems not to be in sync with main. This makes testing a bit difficult.

LiangJYu commented:
1. `dual-pol` works for me with my runconfig and dataset.
2. Fixed - missed this detail from `create_output_paths`
3. Fixed - somehow `git merge main` missed this

vbrancat commented:
@LiangJYu can you share your run configuration file and the generated product or point me to that?

LiangJYu commented:
> @LiangJYu can you share your run configuration file and the generated product or point me to that?

Runconfig and product are on aurora at /home/lyu/data/more_than_one_pol_CSLC

vbrancat reviewed (commented):
I was able to test the PR and when `split_spectrum` is disabled, I do see the multiple polarization processed. However, when I enable split_spectrum with `dual-pol` or `co-pol` processing I get a big fat segmentation fault. I would say that the Segmentation fault is not related to this PR but it should be addressed asap. Let's just open an issue to keep track of that. Will approve once the issue is opened.

LiangJYu commented:
> I was able to test the PR and when `split_spectrum` is disabled, I do see the multiple polarization processed. However, when I enable split_spectrum with `dual-pol` or `co-pol` processing I get a big fat segmentation fault. I would say that the Segmentation fault is not related to this PR but it should be addressed asap. Let's just open an issue to keep track of that. Will approve once the issue is opened.

Issue is up. If you can have the segfault trace handy, can you post it the [issue](https://github.com/opera-adt/COMPASS/issues/86)?

vbrancat reviewed (approved):
The multi-polarization processing part looks good. We will need to address what is happening with the split-spectrum. Yay

vbrancat commented:
@scottstanie do you have time to review this? Thanks :)

scottstanie reviewed (approved):
We should probably remove the `_temp.vrt` file after using it.

```bash
$ ls t078_165496_iw2/20220713/
t078_165496_iw2_20220713.h5  t078_165496_iw2_20220713_VH_temp.vrt
```


But i believe this ran fine for me:

```
$ h5ls t078_165496_iw2/20220713/t078_165496_iw2_20220713.h5/science/SENTINEL1/CSLC/grids
VH                       Dataset {4545, 20200}
VV                       Dataset {4545, 20200}
projection               Dataset {SCALAR}
x_coordinates            Dataset {20200}
x_spacing                Dataset {SCALAR}
y_coordinates            Dataset {4545}
y_spacing                Dataset {SCALAR}
```

LiangJYu commented on src/compass/utils/runconfig.py:
```diff
@@ -290,14 +290,16 @@ def create_output_paths(sns, bursts):
         os.makedirs(scratch_path, exist_ok=True)
 
         output_paths[path_key] = SimpleNamespace(output_directory=out_dir,
                                                  file_name_stem=fname_stem,
```
It currently is unused. Could be useful later for output where polarization doesn't matter. I will remove it if you want it gone.

scottstanie replied:
ok that's true, not a problem leaving it

scottstanie commented:
@LiangJYu oh ok, yes for that one I put them as the same folder 👍

LiangJYu commented on src/compass/s1_geocode_slc.py:
```diff
@@ -139,28 +124,58 @@ def run(cfg: GeoRunConfig):
+                else:
+                    rdr_burst_raster = isce3.io.Raster(temp_slc_path)
+
+                init_geocoded_dataset(grid_group, pol, geo_grid, 'complex64',
+                                      f'{pol} geocoded CSLC image')
+
+                # access the HDF5 dataset for a given frequency and polarization
+                dataset_path = f'{grid_path}/{pol}'
```
@scottstanie would you be ok with me not addressing this? The index reuse in geocodeSlc array mode [PR](https://github-fn.jpl.nasa.gov/isce-3/isce/pull/1236) just got merged into ISCE3 develop and that allows for implementation of array mode geocodeSlc which would:
1. not require a raster to be passed to `geocode_slc`
2. allow for a more efficient fix of #86. With geocode_slc raster mode, there's no clean way to take advantage geocode index resuse with multiple output rasters in the same HDF5. VRT of IH5 does not work, so temporary files will be necessary.

scottstanie replied:
ok yep, no problem
